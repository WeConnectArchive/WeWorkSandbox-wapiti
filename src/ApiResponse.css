.title{font-family:var(--font-regular);font-size:var(--title-font-size);font-weight:bold;margin-bottom:8px}.resp-head{vertical-align:middle;padding:16px 0 8px}.resp-head.divider{border-top:1px solid var(--border-color);margin-top:10px}.resp-status{font-weight:bold;font-size:calc(var(--small-font-size) + 1px)}.resp-descr{font-size:calc(var(--small-font-size) + 1px);color:var(--light-fg)}.top-gap{margin-top:16px}.tab-buttons{height:30px;border-bottom:1px solid var(--light-border-color);align-items:stretch}.tab-btn{color:var(--fg);border:none;background-color:transparent;cursor:pointer;padding:1px;outline:none;font-size:var(--small-font-size);margin-right:16px;padding:1px}.tab-btn.active{border-bottom:3px solid var(--primary-color);font-weight:bold;color:var(--primary-color)}.tab-btn:hover{color:var(--primary-color)}.tab-content{margin:-1px 0 0 0}.descr-text{color:var(--light-fg);font-family:var(--font-regular)}.tree{padding:16px 2px}@media only screen and (min-width: 768px){.tree{padding:16px}}
