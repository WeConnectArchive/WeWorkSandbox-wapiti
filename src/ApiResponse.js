import React, {Component} from 'react';
import cx from 'classnames';
import marked from 'marked';
import DangerousHTML from 'react-dangerous-html';
import {schemaToModel, generateExample, removeCircularReferences} from './utils/common-utils.js';
import {JsonTree} from './JsonTree';
import {SchemaTree} from './SchemaTree';

import style from './ApiResponse.scss';
import stylee from './EndPoint.scss';
import stylefo from './FontStyles.scss';
import stylett from './TabStyles.scss';
import stylef from './FlexStyles.scss';

export class ApiResponse extends Component {

  constructor(props) {
    super(props);
    let eta = [];
    for(let statusCode in this.props.responses) {
      eta[statusCode] = true;
    }
    this.state = { exampleTabActive: eta };
  }

  render() {
    return (
    <div className={cx(stylee["response"],stylef["col"], stylefo["regular-font"])}>
    <div className={cx(style["title"])}>RESPONSE</div>
    {this.responseTemplate()}
    </div>
   );
  }

  activateTab = (e) => {
    let id = e.target.id;
    let status = id.split('_')[0];
    let eta = this.state.exampleTabActive;
    eta[status] = id.endsWith('_example'); 
    this.setState({ exampleTabActive: eta });
  }

  responseTemplate(){

    let selectedMimeValueForEachStatus={};
    let headersForEachRespStatus={};
    let mimeResponsesForEachStatus={};

    for(let statusCode in this.props.responses) {
      let allMimeResp={};
      if (this.props.specType === 'grpc') {
          let respSchemaTree =  this.props.responses[statusCode].resp.schema;
          allMimeResp['protobuf'] = {
            "examples"  : [],
            "schemaTree": respSchemaTree
          }
          allMimeResp['protobuf'].examples.push({exampleValue: this.props.responses[statusCode].resp.mock});
      }
      for(let mimeResp in this.props.responses[statusCode].content ) {
        let mimeRespObj = this.props.responses[statusCode].content[mimeResp];
        {/*Remove Circular references from Response schema*/}
        try {
          mimeRespObj.schema = JSON.parse(JSON.stringify(mimeRespObj.schema, removeCircularReferences(0)));
        }
        catch (err) {
          console.error("Unable to resolve circular refs in schema", mimeRespObj.schema);
          return;
        }

        // Generate Schema
        let schemaTree = schemaToModel(mimeRespObj.schema,{});

        // Generate Example
        let respExample = generateExample(
          mimeRespObj.schema? mimeRespObj.schema.examples:'',
          mimeRespObj.schema? mimeRespObj.schema.example:'',
          mimeRespObj.schema,
          mimeResp,
          "json"
        );
        allMimeResp[mimeResp] = {
          "description":this.props.responses[statusCode].description,
          "examples"  : respExample,
          "schemaTree": schemaTree,
        }
        selectedMimeValueForEachStatus[statusCode]= mimeResp;
      }

      // Headers for each response status
      let tempHeaders = [];
      for (let key in this.props.responses[statusCode].headers){
        tempHeaders.push ( { "name":key, ...this.props.responses[statusCode].headers[key]} );
      }
      headersForEachRespStatus[statusCode]   = tempHeaders;
      mimeResponsesForEachStatus[statusCode] = allMimeResp;
    } 

    let styleActive = { flex: 1 }
    let styleNotActive = { flex: 1, display: 'none' }

    return (<React.Fragment>{Object.keys(this.props.responses).map(
      (status, index)  => (<React.Fragment>
      <div className={cx(style["resp-head"], style[(index===0?'top-gap':'divider')])}><React.Fragment>
        {this.props.specType === 'grpc'? <span className={cx(style["resp-status"])}>{this.props.responses[status].resp.objType}</span>  : <span className="resp-status">{status}:</span>}
        <span className={cx(style["resp-descr"])}>{this.props.responses[status].description}</span>
        { (headersForEachRespStatus[status] && headersForEachRespStatus[status].length > 0)? (
          <React.Fragment><div style={{ padding: '12px 0 5px 0' }} className={cx(style["resp-status"])}>Response Headers:</div>
          <table style={{}}>
            {headersForEachRespStatus[status].map( v => (
              <tr>
                <td style={{ padding: '0 12px', verticalAlign: 'top' }} className={cx(stylefo["regular-font-size"])}> {v.name}</td>
                <td style={{ padding: '0 12px', verticalAlign: 'top', lineHeight: '14px' }} className={cx(style["descr-text"],stylefo["small-font-size"])}>
                  <span className={cx(stylefo["m-markdown-small"])}><DangerousHTML html={marked(v.description)} /></span>
                  { (v.schema && v.schema.example)? (<div><br/><span style={{ fontWeight: 'bold' }}>EXAMPLE:</span> {v.schema.example}</div>):'' }
                </td>
              </tr>
            ))}
          </table>
         </React.Fragment>):''}
        </React.Fragment>
      </div>
      {Object.keys(mimeResponsesForEachStatus[status]).map(
        mimeType => mimeType.includes('octet-stream')? (<React.Fragment> <span style={{ color: 'var(--primary-color)' }}> Content-Type: </span> {mimeType} (Binary Data) </React.Fragment>): (
          <div className={cx(stylett["tab-panel"], stylef["col"])} style={{ borderWidth: '0',  minHeight: '200px' }}>
            <div id={status+'_'+mimeType+'_tab-buttons'} onClick={ this.activateTab } className={cx(stylett["tab-buttons"],stylef["row"])} >
              <button className={cx(style['tab-btn'],style[(this.state.exampleTabActive[status]?'active':'')])} id={status+'_'+mimeType+'_example'}>EXAMPLE</button>
              <button className={cx(style['tab-btn'],style[(this.state.exampleTabActive[status]?'':'active')])} id={status+'_'+mimeType+'_model'}>MODEL</button>
              <div style={{ flex:1 }}></div>
              <div style={{ alignSelf: 'center', fontSize: 'var(--small-font-size)' }}> {mimeType} </div>
            </div>
            <div id={status+'_'+mimeType+'_example'} className={cx(stylett["tab-content"], stylef["col"])} style={this.state.exampleTabActive[status]?styleActive:styleNotActive}>
              <JsonTree data={mimeResponsesForEachStatus[status][mimeType].examples[0].exampleValue}></JsonTree>
            </div>
            <div id={status+'_'+mimeType+'_model'} className={cx(stylett["tab-content"], stylef["col"])} style={this.state.exampleTabActive[status]?styleNotActive:styleActive}>
              <SchemaTree data={mimeResponsesForEachStatus[status][mimeType].schemaTree}></SchemaTree>
            </div>
          </div>)
      )}</React.Fragment>)
    )}</React.Fragment>)

  }
}
