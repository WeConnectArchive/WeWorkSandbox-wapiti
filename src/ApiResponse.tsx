import * as React from 'react';
import marked from 'marked';
import DangerousHTML from 'react-dangerous-html';
import {schemaToModel, generateExample, removeCircularReferences} from './utils/common-utils.js';
import {JsonTree} from './JsonTree';
import {SchemaTree} from './SchemaTree';
import './ApiResponse.css'
  
interface ApiResponseProps {
    specType: string;
    responses: any;
};

interface ApiResponseState {
    exampleTabActive: boolean[];
};

export class ApiResponse extends React.Component<ApiResponseProps, ApiResponseState> {

  constructor(props: ApiResponseProps) {
    super(props);
    let eta = [];
    for(let statusCode in this.props.responses) {
      eta[statusCode] = true;
    }
    this.state = { exampleTabActive: eta };
  }

  render() {
    return (
    <div className="response col regular-font">
    <div className="title">RESPONSE</div>
    {this.responseTemplate()}
    </div>
   );
  }

  activateTab = (e) => {
    let id = e.target.id;
    let status = id.split('_')[0];
    let eta = this.state.exampleTabActive;
    eta[status] = id.endsWith('_example'); 
    this.setState({ exampleTabActive: eta });
  }

  responseTemplate(){

    let selectedMimeValueForEachStatus={};
    let headersForEachRespStatus={};
    let mimeResponsesForEachStatus={};

    for(let statusCode in this.props.responses) {
      let allMimeResp={};
      if (this.props.specType === 'grpc') {
          let respSchemaTree =  this.props.responses[statusCode].resp.schema;
          allMimeResp['protobuf'] = {
            "examples"  : [],
            "schemaTree": respSchemaTree
          }
          allMimeResp['protobuf'].examples.push({exampleValue: this.props.responses[statusCode].resp.mock});
      }
      for(let mimeResp in this.props.responses[statusCode].content ) {
        let mimeRespObj = this.props.responses[statusCode].content[mimeResp];
        {/*Remove Circular references from Response schema*/}
        try {
          mimeRespObj.schema = JSON.parse(JSON.stringify(mimeRespObj.schema, removeCircularReferences(0)));
        }
        catch{
          console.error("Unable to resolve circular refs in schema", mimeRespObj.schema);
          return;
        }

        // Generate Schema
        let schemaTree = schemaToModel(mimeRespObj.schema,{});

        // Generate Example
        let respExample = generateExample(
          mimeRespObj.schema? mimeRespObj.schema.examples:'',
          mimeRespObj.schema? mimeRespObj.schema.example:'',
          mimeRespObj.schema,
          mimeResp,
          "json"
        );
        allMimeResp[mimeResp] = {
          "description":this.props.responses[statusCode].description,
          "examples"  : respExample,
          "schemaTree": schemaTree,
        }
        selectedMimeValueForEachStatus[statusCode]= mimeResp;
      }

      // Headers for each response status
      let tempHeaders:any[] = [];
      for (let key in this.props.responses[statusCode].headers){
        tempHeaders.push ( { "name":key, ...this.props.responses[statusCode].headers[key]} );
      }
      headersForEachRespStatus[statusCode]   = tempHeaders;
      mimeResponsesForEachStatus[statusCode] = allMimeResp;
    } 

    let styleActive = { flex: 1 }
    let styleNotActive = { flex: 1, display: 'none' }

    return (<>{Object.keys(this.props.responses).map(
      (status, index)  => (<>
      <div className="resp-head {index===0?'top-gap':'divider'}"><>
        {this.props.specType === 'grpc'? <span className="resp-status">{this.props.responses[status].resp.objType}</span>  : <span className="resp-status">{status}:</span>}
        <span className="resp-descr">{this.props.responses[status].description}</span>
        { (headersForEachRespStatus[status] && headersForEachRespStatus[status].length > 0)? (
          <><div style={{ padding: '12px 0 5px 0' }} className="resp-status">Response Headers:</div>
          <table style={{}}>
            {headersForEachRespStatus[status].map( v => (
              <tr>
                <td style={{ padding: '0 12px', verticalAlign: 'top' }} className="regular-font-size"> {v.name}</td>
                <td style={{ padding: '0 12px', verticalAlign: 'top', lineHeight: '14px' }} className="descr-text small-font-size">
                  <span className="m-markdown-small"><DangerousHTML html={marked(v.description)} /></span>
                  { (v.schema && v.schema.example)? (<><br/><span style={{ fontWeight: 'bold' }}>EXAMPLE:</span> {v.schema.example}</>):'' }
                </td>
              </tr>
            ))}
          </table>
         </>):''}
        </>
      </div>
      {Object.keys(mimeResponsesForEachStatus[status]).map(
        mimeType => mimeType.includes('octet-stream')? (<div> <span style={{ color: 'var(--primary-color)' }}> Content-Type: </span> {mimeType} (Binary Data) </div>): (
          <div className="tab-panel col" style={{ borderWidth: '0',  minHeight: '200px' }}>
            <div id={status+'_'+mimeType+'_tab-buttons'} onClick={ this.activateTab } className="tab-buttons row" >
              <button className={'tab-btn'+(this.state.exampleTabActive[status]?' active':'')} id={status+'_'+mimeType+'_example'}>EXAMPLE</button>
              <button className={'tab-btn'+(this.state.exampleTabActive[status]?'':' active')} id={status+'_'+mimeType+'_model'}>MODEL</button>
              <div style={{ flex:1 }}></div>
              <div style={{ alignSelf: 'center', fontSize: 'var(--small-font-size)' }}> {mimeType} </div>
            </div>
            <div id={status+'_'+mimeType+'_example'} className="tab-content col" style={this.state.exampleTabActive[status]?styleActive:styleNotActive}>
              <JsonTree data={mimeResponsesForEachStatus[status][mimeType].examples[0].exampleValue}></JsonTree>
            </div>
            <div id={status+'_'+mimeType+'_model'} className="tab-content col" style={this.state.exampleTabActive[status]?styleNotActive:styleActive}>
              <SchemaTree data={mimeResponsesForEachStatus[status][mimeType].schemaTree}></SchemaTree>
            </div>
          </div>)
      )}</>)
    )}</>)

  }
}
