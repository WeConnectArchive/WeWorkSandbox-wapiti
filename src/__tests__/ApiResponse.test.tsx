import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { ApiResponse } from '../ApiResponse';

const theResponses = {
  "200": {
      description: "description",
      content: {
        "application/json": {
           schema:{
             type: "object",
             required: ["id","name"],
             properties: {
               id: {
                 type: "integer",
                 format: "int64"
               },
               name: {
                 type: "string"
               },
               tag: {
                 type: "string"
               }
             }
           }
        }
      }
   }
}

describe('ApiResponse', () => {
  it('renders with rest', () => {
    let wrapper = shallow(<ApiResponse
        responses={theResponses}
        specType="rest"
      ></ApiResponse>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

});
