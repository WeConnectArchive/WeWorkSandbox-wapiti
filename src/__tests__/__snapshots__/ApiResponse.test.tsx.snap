// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiResponse renders with rest 1`] = `
<div
  className="response col regular-font"
>
  <div
    className="title"
  >
    RESPONSE
  </div>
  <div
    className="resp-head {index===0?'top-gap':'divider'}"
  >
    <span
      className="resp-status"
    >
      200
      :
    </span>
    <span
      className="resp-descr"
    >
      description
    </span>
  </div>
  <div
    className="tab-panel col"
    style={
      Object {
        "borderWidth": "0",
        "minHeight": "200px",
      }
    }
  >
    <div
      className="tab-buttons row"
      id="200_application/json_tab-buttons"
      onClick={[Function]}
    >
      <button
        className="tab-btn active"
        id="200_application/json_example"
      >
        EXAMPLE
      </button>
      <button
        className="tab-btn"
        id="200_application/json_model"
      >
        MODEL
      </button>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <div
        style={
          Object {
            "alignSelf": "center",
            "fontSize": "var(--small-font-size)",
          }
        }
      >
         
        application/json
         
      </div>
    </div>
    <div
      className="tab-content col"
      id="200_application/json_example"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <JsonTree
        data={
          Object {
            "id": 0,
            "name": "string",
            "tag": "string",
          }
        }
      />
    </div>
    <div
      className="tab-content col"
      id="200_application/json_model"
      style={
        Object {
          "display": "none",
          "flex": 1,
        }
      }
    >
      <SchemaTree
        data={
          Object {
            "id*": "integerÂ int64~|~",
            "name*": "string~|~",
            "tag": "string~|~",
          }
        }
      />
    </div>
  </div>
</div>
`;
